### PR TITLE
MRI Segmentation 3D U-Net

### PR DESCRIPTION
**Pull request by elliana-r**
This pull request introduces a 3D U-Net model for multiclass segmentation, targeting six classes. The model has been trained and evaluated on a relevant dataset, and its performance has been tracked through Dice scores for each class and loss metrics for both training and validation sets.

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**
Recognition problem: total 15 Marks
Problem solving: (5)
•	Algorithm was able solve the problem with Dice score above 0.71 for each class for the testing set. 
Algorithm implementation: (3)
•	Algorithm implemented in PyTorch for the UNet3D model and functions.
Design: Good coding design (1)
Commenting: Sufficient comments throughout the code (1)
Difficulty: 3D UNet – Normal (5)

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. UNet3D (Normal Difficulty)</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Late submission.</td><td>-2</td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove model checkpoint for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-10</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by hanemma7moud** [comment]
Elliana Rogaczewski-Novak

**Comment by elliana-r** [comment]
> Feedback marks possible +2 if the requested changes are made (see above).

Based on the feedback, I have added header blocks and removed the model checkpoint.
Thank you.

**Comment by shakes76** [comment]
Approved extension +2, feedback applied +2, Segmentations not one hot encoded -2 additionally
